File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/IfConditionalAlwaysTrueOrFalse.cs
using System;

namespace Tests.Diagnostics
{
    public class IfConditionalAlwaysTrueOrFalse
    {
        public void DoSomething() { throw new NotSupportedException(); }
        public void DoSomething2() { throw new NotSupportedException(); }
        public IfConditionalAlwaysTrueOrFalse(bool a, bool b)
        {
            var someWronglyFormatted =      45     ;
            if (a == b)
            {
                DoSomething();
            }

            if (true == b)
            {
                DoSomething2();
            }

            if (a)
            {
                DoSomething();
            }

            if (true) // Noncompliant {{Remove this useless 'if' statement.}}
//          ^^^^^^^^^
            {
                DoSomething2();
            }

            if (false) // Noncompliant {{Remove this useless 'if' statement.}}
            {
                DoSomething();
            }

            if (true) // Noncompliant
            {
                DoSomething();
            }
            else // Noncompliant {{Remove this useless 'else' clause.}}
            {
            }

            if (false) // Noncompliant
            {
                DoSomething2();
            }
            else
            {
                ; ;
            }
        }
    }
}

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/IfConditionalAlwaysTrueOrFalse.vb
Public Class IfConditionalAlwaysTrueOrFalse

  Public Sub Rspec()
    If True Then ' Noncompliant {{Remove this useless 'If' statement.}}
'   ^^^^^^^^^^^^
      DoSomething()
    Else ' Noncompliant {{Remove this useless 'Else' clause.}}
    End If

    If False Then ' Noncompliant
'   ^^^^^^^^^^^^^
      DoSomethingElse()
    End If

    If 2 < 3 Then ' FN
    End If

    Dim i As Integer = 0
    Dim j As Integer = 0
    j = Foo()

    If j > 0 AndAlso i > 0 Then ' FN
      ' ...
    End If

    Dim b As Boolean = True
    If b OrElse Not b Then ' FN
      ' ...
    End If

    If j > 1 AndAlso True Then
    End If

    If j > 1 OrElse True Then ' FN
    End If

    If j > 1 AndAlso False Then ' FN
    End If

    If j > 1 OrElse True Then
    End If

    If True Then ' Noncompliant
      DoSomething()
    ElseIf True Then ' Noncompliant {{Remove this useless 'ElseIf' clause.}}
      DoSomething()
    ElseIf False Then ' Noncompliant {{Remove this useless 'ElseIf' clause.}}
      DoSomething()
    Else ' Noncompliant {{Remove this useless 'Else' clause.}}
      DoSomething()
    End If

    If False Then ' Noncompliant
      DoSomethingElse()
    ElseIf True Then
      DoSomething()
    Else
      DoSomething()
    End If

  End Sub

  Public Sub DoSomething()
  End Sub

  Public Sub DoSomethingElse()
  End Sub

  Public Function Foo() As Integer
    Return 42
  End Function

  Public Sub New(ByVal a As Boolean, ByVal b As Boolean)
    Dim someWronglyFormatted = 45

    If a = b Then
      DoSomething()
    End If

    If True = b Then
      DoSomethingElse()
    End If

    If a Then
      DoSomething()
    End If
  End Sub

End Class
